### PR TITLE
Provide

### PR DESCRIPTION
**Pull request by eddow**
Add the @Provide decorator - the properties or property-get that are such decorated are evaluated on provision, added to the previous "provide" object or "provide" function-return-value if there was any

### CODE REVIEW

**Comment by kaorun343** [comment]
@eddow 

Hi.

Please write a test.

File: src/vue-property-decorator.ts [hunk 1]
```diff
@@ -26,13 +26,36 @@ export function Inject(key?: string | symbol): PropertyDecorator {
...
+  })
+}
+
+/**
  * decorator of model
  * @param  event event name
  * @return PropertyDecorator
  */
-export function Model(event: string): PropertyDecorator {
+export function Model(event?: string): PropertyDecorator {
```
Comment (kaorun343) [inline src/vue-property-decorator.ts]: This change looks good to me.
However, is this change related to `@Provide`?

File: src/vue-property-decorator.ts [hunk 2]
```diff
@@ -26,13 +26,36 @@ export function Inject(key?: string | symbol): PropertyDecorator {
...
+ * @return PropertyDecorator | void
+ */
+export function Provide(key?: string | symbol): PropertyDecorator {
+  return createDecorator((componentOptions, k) => {
+		var provide: any = componentOptions.provide;
+		if('function'!== typeof provide || !provide.managed) {
+			var original = provide;
+			provide = componentOptions.provide = {
+				provide() {
+					var i, rv = Object.create(('function'=== typeof original?original.call(this):original)||null);
```
Comment (kaorun343) [inline src/vue-property-decorator.ts]: `typeof original === 'function' will always return `false`, I thought.
Also, could you tell me what the name of `rv` implies?

File: src/vue-property-decorator.ts (hunk 1 above)
Comment (eddow) [inline src/vue-property-decorator.ts]: Oh yhea, sorry, I did it for my personal use and it slipped in the PR...

File: test/decorator.spec.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import * as Vue from 'vue'
-import { Component, Inject, Model, Prop, Watch } from '../src/vue-property-decorator'
+import { Provide, Component, Inject, Model, Prop, Watch } from '../src/vue-property-decorator'
```
Comment (kaorun343) [inline test/decorator.spec.ts]: Please sort by name

File: src/vue-property-decorator.ts (hunk 2 above)
Comment (eddow) [inline src/vue-property-decorator.ts]: `rv` stands for return-value; please, feel free to use any name
A scenario like the following would make `original` an object at the first @Provide usage, and is allowed by Vue
`@Component({ provide: { valA: "a-constant" } })`

File: src/vue-property-decorator.ts [hunk 4]
```diff
@@ -26,13 +26,36 @@ export function Inject(key?: string | symbol): PropertyDecorator {
...
 
 /**
+ * decorator of a provide
+ * @param key key
+ * @return PropertyDecorator | void
+ */
+export function Provide(key?: string | symbol): PropertyDecorator {
+  return createDecorator((componentOptions, k) => {
+		var provide: any = componentOptions.provide;
+		if('function'!== typeof provide || !provide.managed) {
```
Comment (kaorun343) [inline src/vue-property-decorator.ts]: I couldn't understand how `componentOptions.provide.managed` works...
Instead of adding a new property, I recommend using `WeakMap` to manage something you'd like to do.

```ts
let container: WeakMap<ComponentOptions<Vue>, object> | null = null

export function Provide(key?: string | symbol): PropertyDecorator {
  // for the browsers doesn't support WeakMap
  // Both WeakMap and Symbol are ES2015 features.
  if (container == null) container = new WeakMap()
  return createDecorator((componentOptions, k) => {
    const managed = container.get(componentOptions)
    // ...
  }
}
```

File: src/vue-property-decorator.ts (hunk 1 above)
Comment (kaorun343) [inline src/vue-property-decorator.ts]: You don't have to change this section 👍

Comment (kaorun343) [inline src/vue-property-decorator.ts]: Ah, you don't have to revert this section.

File: src/vue-property-decorator.ts (hunk 4 above)
Comment (eddow) [inline src/vue-property-decorator.ts]: `managed` is the list of the "managed provides" (the ones managed by this library) - it is a map {k: v} where `v` is the name used in the "provide" list, and `k` is the property of the instance that is queried when providing.
Everything is a string here - I don't see a use for a WeakMap

File: src/vue-property-decorator.ts (hunk 2 above)
Comment (kaorun343) [inline src/vue-property-decorator.ts]: I understand.

**Comment by eddow** [comment]
Don't forget to publish it on npm - I try to clean my project folders ;-)
